### ISSUE TITLE
Displaying a non-primitive object via VBGeval does not show all content (GDB).

### ISSUE THREAD

**Issue by ozankabak** [idanarye/vim-vebugger#25]
I am seeing some unexpected behavior while using GDB via Vebugger. Consider the following type:

```
struct B {
   int foo, foobar, foobarfoo, foobarbaz;
};
```

Assume I initialize an instance `b` of type `B`, with all its fields set to `3`. When I try to print this instance `b` via `VBGeval`, I get:

```
b: {foobarfoo = 3
```

Obviously, I expect to see:

```
b: {foobarfoo = 3, foobar = 3, foo = 3, foobarbaz = 3}
```

**Comment by idanarye** [comment]
The full interaction with GDB looks something like this:

```
&"print b\n"
~"$1 = {foo = 3"
~", foobar = 3"
~", foobarfoo = 3"
~", foobarbaz = 3"
~"}"
~"\n"
^done
```

You are only seeing the first line of the response(which for some reason is not `foo` in your case, even though it's the first field of `B`...). I might need some architectural changes to aggregate all the lines of the response...

**Comment by ozankabak** [comment]
Ah, I realized I changed the variable order at some point. The response I actually get is `b: {foo = 3`, which is only the first line of the full response, as you said.

When I use `VBGrawWrite print b`, and inspect the terminal buffer, I indeed see the entire response. So, `VBGeval` seems to print only the first line of the response. I think it is more intuitive for `VBGeval` to print the full response. Would that be a hard-to-make change? If you agree that this change makes sense, but do not have much time, I am willing to help if you can give me some pointers.

**Comment by idanarye** [comment]
This requires a bit of design. Vebugger's architecture uses what I call _Read Handlers_ to divine the meaning of the debugger's output. These read handlers receive the debugger output line by line and parse each line separately, so it's easy to identify the `$1 =` part in the first line - but the next lines are a bit trickier.

The simplest solution will be to modify `vebugger#gdb#_readEvaluatedExpressions` to keep the lines it gets until it gets `~"\n"` - but I'm not sure I'm fine with this fix. I don't think it's the read handlers' job to aggregate lines like this. If it makes sense to aggregate lines, it makes sense to do so for all read handlers...

What I have in mind is to allow debugger implementations to set their own debugger output aggregation, instead of the current hard-coded line-by-line aggregation. But this is something I should probably do myself, since it's an architectural change.

**Comment by ozankabak** [comment]
Well... I don't think using `'\n'` for delimiting makes sense in general. After all, debuggers routinely output multiple-line responses to many commands. In other words, a single output might contain multiple `'\n'` characters. Consider the `explore b` command, which elicits the following response:

```
explore b
&"explore b\n"
~"The value of 'b' is a struct/class of type 'B' with the following fields:\n"
~"\n"
~"        foo = 3 .. (Value of type 'int')\n"
~"     foobar = 3 .. (Value of type 'int')\n"
~"  foobarfoo = 3 .. (Value of type 'int')\n"
~"  foobarbaz = 3 .. (Value of type 'int')\n"
~"\n"
^done
```

Aggregating based on `'\n'` would still fail in this case, no?

Here is my suggestion: Obviously, Vebugger already knows where the output finishes. So, shouldn't it (by default) collect all lines together, until the output is consumed, and parse the entire thing?

**Comment by idanarye** [comment]
> Obviously, Vebugger already knows where the output finishes.

That's the problem - it doesn't. It supports multiple debuggers, each with it's own way of indicating "end of output", but usually single answers were contained in single lines, so I did not implement a method for divining the end of input for specific debuggers.

I may need to be smarter about this though. For user commands, it seems like they begin with `&"command\n"`, then many lines that look like `~"output"`, and finally `^done`. This should be pretty easy.

It can get more complicated for output from things like breakpoints or `display`. For example, I've set `display` on two `B`s - `a` and `b` - and `continue`d until a breakpoint. I got this:

```
&"continue\n"
~"Continuing.\n"
^running
*running,thread-id="all"
(gdb) 
=breakpoint-modified,bkpt={number="3",type="breakpoint",disp="keep",enabled="y",addr="0x0000000000400686",func="main(int, char**)",file="tests/gdb/test.cpp",fullname="/media/d/Programming/Projects/vim-vebugger/tests/gdb/test.cpp",line="48",thread-groups=["i1"],times="1",original-location="tests/gdb/test.cpp:48"}
~"2: b = {foo = 3"
~", foobar = 3"
~", foobarfoo = 3"
~", foobarbaz = 3"
~"}"
~"\n"
~"1: a = {foo = 2"
~", foobar = 2"
~", foobarfoo = 2"
~", foobarbaz = 2"
~"}"
~"\n"
~"\nBreakpoint "
~"3, main (argc=1, argv=0x7fffffffe648) at tests/gdb/test.cpp:48\n"
~"48\t    printf(\"4\");\n"
~"2: b = {foo = 3"
~", foobar = 3"
~", foobarfoo = 3"
~", foobarbaz = 3"
~"}"
~"\n"
~"1: a = {foo = 2"
~", foobar = 2"
~", foobarfoo = 2"
~", foobarbaz = 2"
~"}"
~"\n"
*stopped,frame={level="0",func="main",args=[{name="argc",value="1"},{name="argv",value="0x7fffffffe648"}],file="tests/gdb/test.cpp",fullname="/media/d/Programming/Projects/vim-vebugger/tests/gdb/test.cpp",line="48"},reason="breakpoint-hit",disp="keep",bkptno="3",frame={addr="0x0000000000400686",func="main",args=[{name="argc",value="1"},{name="argv",value="0x7fffffffe648"}],file="tests/gdb/test.cpp",fullname="/media/d/Programming/Projects/vim-vebugger/tests/gdb/test.cpp",line="48"},thread-id="1",stopped-threads="all",core="3"
```

In this case, the only way to make sense out of it is to use the `"\n"`s - and then we get this:

```
& continue
~ Continuing.
^ running
* running,thread-id="all"
= breakpoint-modified,bkpt={number="3",type="breakpoint",disp="keep",enabled="y",addr="0x0000000000400686",func="main(int, char**)",file="tests/gdb/test.cpp",fullname="/media/d/Programming/Projects/vim-vebugger/tests/gdb/test.cpp",line="48",thread-groups=["i1"],times="1",original-location="tests/gdb/test.cpp:48"}
~ 2: b = {foo = 3, foobar = 3, foobarfoo = 3, foobarbaz = 3}
~ 1: a = {foo = 2, foobar = 2, foobarfoo = 2, foobarbaz = 2}
~
~ Breakpoint 3, main (argc=1, argv=0x7fffffffe648) at tests/gdb/test.cpp:48
~ 48        printf("4")
~ 2: b = {foo = 3, foobar = 3, foobarfoo = 3, foobarbaz = 3}
~ 1: a = {foo = 2, foobar = 2, foobarfoo = 2, foobarbaz = 2}
* stopped,frame={level="0",func="main",args=[{name="argc",value="1"},{name="argv",value="0x7fffffffe648"}],file="tests/gdb/test.cpp",fullname="/media/d/Programming/Projects/vim-vebugger/tests/gdb/test.cpp",line="48"},reason="breakpoint-hit",disp="keep",bkptno="3",frame={addr="0x0000000000400686",func="main",args=[{name="argc",value="1"},{name="argv",value="0x7fffffffe648"}],file="tests/gdb/test.cpp",fullname="/media/d/Programming/Projects/vim-vebugger/tests/gdb/test.cpp",line="48"},thread-id="1",stopped-threads="all",core="3"
```

Much more parsable, and I think this is how GDB's MI mode is intended to be used.

**Comment by ozankabak** [comment]
I see. Continuously-outputting (i.e. streaming) commands pose a challenge because their outputs are not explicitly delimited.

I agree that output should always be aggregated based on explicit newline markers. However, for synchronous commands, there needs to be another layer of aggregation, which can be done using the beginning/end markers you mentioned.

Such a setup should address all the use cases except continuously-outputting (i.e. streaming) multi-line blocks. What do you think?

**Comment by idanarye** [comment]
Vebugger is using a message bus architecture. It reads output lines from the debugger, and feeds them to the _Read Handlers_. The read handlers update a _Read Result Object_, which is then passed to the _Think Handlers_. The think handlers look at the read result object, interact with the Vim environment(e.g. print data to the user), and set the _Write Actions_. Finally, the _Write Handlers_ read the write actions, and write commands into the debugger.

The handlers within each group(read, think ,write) - each has it's own domain to handle - and while there is interaction between the groups, an handler in one group does not care with which handler in the other group it interacts(e.g. think handlers do not care which read handler created the field in the read result object they are looking at). This is an important quality of Vebugger, which allows it to be extended to support both new features and new debuggers.

To preserve this behavior, I want the parsing of the debugger output to be in the read handlers, but the feature I need to add to fix this bug requires a "preprocessing" step before the read handlers, a bufferer that will convert GDB's MI output to tagged lines. I would like to keep this preprocessing simple and low-level, because
1. I can only have one bufferer per debugger. Read handlers I can have as many as I want, so I prefer to let them handle complex patterns because each read handler only needs to parse one pattern, but with the bufferer each and every detail I need to consider will directly contribute to the overall complicity of the bufferer.
2. I want the bufferer to work directly with the debugger's output streams. It's not going to be an extra state in the debugger output processing pipeline - it's going to replace an existing, "hard-coded"(=still VimScript, but not one that can be set by debugger implementations) bufferer that converts the byte stream to lines. Since it's input is in a low-level format, I don't want it's output format to be too high-level.

So, parsing the MI to tagged lines is a simple-yet-helpful conversion that I want to do. Partitioning these lines based on commands? Not so much.

Now, I don't really want the read handlers to bunch these lines together - but they don't have to! After the basic buffering I want to add, one-line-at-a-time is perfectly fine for any read handler I've implemented or thought about implementing. The ones that are a bit hard to parse line-by-line are the ones that are not direct responses to commands, like getting the stack trace the debugger is ordered to print every time it stops, and these are easily doable with read-handler-state or think-handler-state.

As for your `explore` example - consider this:

```
struct Foo {
   int x;
   int y;
};

struct Bar {
   struct Foo a;
   struct Foo b;
};

int main(int ARGC, char** ARGV) {
    struct Bar bar;

    return 0;
}
```

Now,  let's `explore bar`:

```
&"explore bar\n"
~"The value of 'bar' is a struct/class of type 'struct Bar' with the following fields:\n"
~"\n"
~"  a = <Enter 0 to explore this field of type 'struct Foo'>\n"
~"  b = <Enter 1 to explore this field of type 'struct Foo'>\n"
~"\n"
~"Enter the field number of choice: "
```

If I implemented the bufferer to bunch all the lines resulted from a single command before passing them on, we would be stuck here. It would have identified the start of a command thanks to `&"explore bar\n"`, but there is no end of output(Line starting with `^`) so it won't pass anything anymore to the read handlers!

BTW - this won't work with my MI-to-tagged-lines bufferer either, because there is no `\n` after "_Enter the field number of choice: _". So if I ever want to support `explore`, I'll have to do more drastic changes to the architecture, which I'm not planning to do any time soon...

At any rate, I do plan to do something with the `^done` marker - when the user types a raw command, I can use it to display it's output entirely. But this kind of aggregation I'll probably do in a read handler. It's OK to do it in a read handler, because I don't have to pass it to another read handler that will pass the full response - something which is impossible in the current architecture. I'll just pass it directly to the think handlers, which will print that response to the user without knowing what it means.

**Comment by ozankabak** [comment]
All right. Looking forward to seeing these changes in action. If needed, I will gladly test them for you once committed. Thanks for making Vebugger, I find it quite useful.

**Comment by idanarye** [comment]
@wanderer2 I've pushed a fix. Make sure it works for you.

**Comment by ozankabak** [comment]
Works great. Thanks!

**Comment by ozankabak** [comment]
Hello Idan. I just realized that the old behavior persists in one of my setups. At work, where I have VIM 7.3 (with patches 1-315) and GDB 7.9.1, I still get the old behavior. At home, where I verified your fix, I have VIM 7.4 (with patches 1-1301) and GDB 7.10.1. Also, work binaries are i686 binaries, while home uses x86_64.

I have a feeling that this has to do with the version difference in VIMs and not GDB's (since both are fairly new) or the i686/x86_64 thing. Maybe your fix uses some Vimscript functionality that does not work in older versions?

**Comment by idanarye** [comment]
If this was a Vim version problem, you would probably have gotten a Vim error message(like in #5). I've just built Vim 7.3.315 and checked and it works just fine.

It makes more sense that GDB 7.9.1 formats `struct`s differently than GDB 7.10.1. With Vebugger running with the older GDB, could you please open the terminal buffer with `:VBGtoggleTerminalBuffer`, `:VBGeval` your variable, and copy-paste here the debugger output from the terminal buffer?

**Comment by ozankabak** [comment]
This is what I see:

```
print a
&print a
~$1 = {
~  foo = -134358008,
~  foobar = -135693120,
~  foobarfoo = 0,
~  foobarbaz = -135696396,
~}
^done
(gdb)
```

There is a small difference between this and the newer GDB's output: This first field, `foo`, is displayed in the next line after the `{` delimiter. Do you think this is the breaking change?

**Comment by idanarye** [comment]
Yes... oh, supporting this is going to be nasty...

**Comment by ozankabak** [comment]
Thanks for working with me and putting time into this. Much appreciated.

**Comment by idanarye** [comment]
I've looked through the GDB manual and found [that this is a setting](http://ftp.gnu.org/old-gnu/Manuals/gdb/html_chapter/gdb_9.html#SEC58):

> set print pretty on
>     Cause GDB to print structures in an indented format with one member per line, like this:
> 
> ```
> $1 = {
>   next = 0x0,
>   flags = {
>     sweet = 1,
>     sour = 1
>   },
>   meat = 0x54 "Pork"
> }
> ```
> 
> set print pretty off
>     Cause GDB to print structures in a compact format, like this:
> 
> ```
> $1 = {next = 0x0, flags = {sweet = 1, sour = 1}, \
> meat = 0x54 "Pork"}
> 
> This is the default format. 
> ```

Since `set print pretty off` this is the default, and that article is about GDB 5, I don't think it's related to the GDB version, but to the configuration in your work environment(`.gdbrc` or something).

At any rate, it's better for Vebugger to set turn this off than to require users to turn this off in their `.gdbrc`. I'm going to go over the list in that manual and see if there are other settings Vebugger should set...

**Comment by idanarye** [comment]
@wanderer2 I've pushed a commit that sets the print format. Try it!
